## DSC478 Winter 2019
## Amy Aumpansub
## Due Date: Mar 12th, 2019

## Assignment 4

### <font color = red>I use Python 3.7.2 to complete this assignment.</font>
### <font color = red>Some functions may not compile because the difference in codes in Python version.</font>
### <font color = red>So, I copied and pasted all functions in this Jupyter Notebook</font>

In [1]:
import numpy as np
import pandas as pd
import os

In [6]:
#Change directory working and import data
os.chdir('/Users/amy_a/Desktop/dsc478/assignment/segmentation_data')


## Problem 1 

## a. Load in the image data matrix (with rows as images and columns as features). Also load in the numeric class labels from the segmentation class file. Using your favorite method (e.g., sklearn's min-max scaler), perform min-max normalization on the data matrix so that each feature is scaled to [0,1] range.

In [7]:
#Import class lebel
Class = pd.read_table('segmentation_classes.txt',header=None)

In [8]:
Class.head(5)

0  1
0  GRASS  0
1  GRASS  0
2  GRASS  0
3  GRASS  0
4  GRASS  0

In [9]:
#Import feature names
Features = pd.read_table('segmentation_names.txt',header=None)

In [10]:
Features

0
0    REGION-CENTROID-COL
1    REGION-CENTROID-ROW
2     REGION-PIXEL-COUNT
3   SHORT-LINE-DENSITY-5
4   SHORT-LINE-DENSITY-2
5             VEDGE-MEAN
6               VEDGE-SD
7             HEDGE-MEAN
8               HEDGE-SD
9         INTENSITY-MEAN
10           RAWRED-MEAN
11          RAWBLUE-MEAN
12         RAWGREEN-MEAN
13            EXRED-MEAN
14           EXBLUE-MEAN
15          EXGREEN-MEAN
16            VALUE-MEAN
17       SATURATION-MEAN
18              HUE-MEAN

In [11]:
Features_list = Features[0].tolist()

In [12]:
#Import image data and add feature names
IF = pd.read_csv('segmentation_data.txt',sep=',', names = Features_list)

In [13]:
IF.head(5)

REGION-CENTROID-COL  REGION-CENTROID-ROW  REGION-PIXEL-COUNT  \
0                110.0                189.0                   9   
1                 86.0                187.0                   9   
2                225.0                244.0                   9   
3                 47.0                232.0                   9   
4                 97.0                186.0                   9   

   SHORT-LINE-DENSITY-5  SHORT-LINE-DENSITY-2  VEDGE-MEAN  VEDGE-SD  \
0                   0.0                   0.0    1.000000  0.666667   
1                   0.0                   0.0    1.111111  0.720082   
2                   0.0                   0.0    3.388889  2.195113   
3                   0.0                   0.0    1.277778  1.254621   
4                   0.0                   0.0    1.166667  0.691215   

   HEDGE-MEAN  HEDGE-SD  INTENSITY-MEAN  RAWRED-MEAN  RAWBLUE-MEAN  \
0    1.222222  1.186342       12.925926    10.888889      9.222222   
1    1.444444  0.750309       13.740741    11.666667     10.333334   
2    3.000000  1.520234       12.259259    10.333334      9.333334   
3    1.000000  0.894427       12.703704    11.000000      9.000000   
4    1.166667  1.005540       15.592592    13.888889     11.777778   

   RAWGREEN-MEAN  EXRED-MEAN  EXBLUE-MEAN  EXGREEN-MEAN  VALUE-MEAN  \
0      18.666668   -6.111111   -11.111111     17.222221   18.666668   
1      19.222221   -6.222222   -10.222222     16.444445   19.222221   
2      17.111110   -5.777778    -8.777778     14.555555   17.111110   
3      18.111110   -5.111111   -11.111111     16.222221   18.111110   
4      21.111110   -5.111111   -11.444445     16.555555   21.111110   

   SATURATION-MEAN  HUE-MEAN  
0         0.508139  1.910864  
1         0.463329  1.941465  
2         0.480149  1.987902  
3         0.500966  1.875362  
4         0.442661  1.863654

In [14]:
IF.shape

(2100, 19)

#### Normarilize data to (0-1) using MinMaxScaler

In [15]:
#Normalize data
from sklearn import preprocessing

In [16]:
min_max_scaler = preprocessing.MinMaxScaler().fit(IF)
IF_norm = min_max_scaler.transform(IF)

/anaconda3/lib/python3.7/site-packages/sklearn/preprocessing/data.py:323: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by MinMaxScaler.
  return self.partial_fit(X, y)


In [17]:
IF_norm[0:5]

array([[4.30830040e-01, 7.41666667e-01, 0.00000000e+00, 0.00000000e+00,
        0.00000000e+00, 3.42205474e-02, 6.72233922e-04, 2.73291926e-02,
        8.55743510e-04, 9.01110284e-02, 7.94165331e-02, 6.11192912e-02,
        1.30943107e-01, 7.31343290e-01, 1.41176540e-02, 8.72865267e-01,
        1.23711348e-01, 5.08138840e-01, 8.31849232e-01],
       [3.35968379e-01, 7.33333333e-01, 0.00000000e+00, 0.00000000e+00,
        0.00000000e+00, 3.80228046e-02, 7.26095734e-04, 3.22981359e-02,
        5.41219947e-04, 9.57913810e-02, 8.50891441e-02, 6.84830672e-02,
        1.34840205e-01, 7.29477615e-01, 2.35294199e-02, 8.59582565e-01,
        1.27393216e-01, 4.63329080e-01, 8.36986460e-01],
       [8.85375494e-01, 9.70833333e-01, 0.00000000e+00, 0.00000000e+00,
        0.00000000e+00, 1.15969577e-01, 2.21344355e-03, 6.70807367e-02,
        1.09658970e-03, 8.54634659e-02, 7.53646732e-02, 6.18556741e-02,
        1.20031165e-01, 7.36940304e-01, 3.88235327e-02, 8.27324481e-01,
        1.13402054e-01

**After mix-max normization, all image data ranges from 0 to 1 as you can see from the array above. It has the min value of 0 and the max value of 1.**

## b. Next, Perform Kmeans clustering (for this problem, use the Kmeans implementation in scikit-learn) on the image data (since there are a total 7 pre-assigned image classes, you should use K = 7 in your clustering). Use Euclidean distance as your distance measure for the clustering. Print the cluster centroids (use some formatting so that they are visually understandable). Compare your 7 clusters to the 7 pre-assigned classes by computing the Completeness and Homogeneity values of the generated clusters.

In [18]:
from sklearn.cluster import KMeans
Kmeans7 = KMeans(n_clusters=7, max_iter=500,verbose =1)
Kmeans7.fit(IF_norm)

Initialization complete
start iteration
done sorting
end inner loop
Iteration 0, inertia 417.63889413405786
start iteration
done sorting
end inner loop
Iteration 1, inertia 393.75278256266245
start iteration
done sorting
end inner loop
Iteration 2, inertia 389.7850015007274
start iteration
done sorting
end inner loop
Iteration 3, inertia 386.9026670225386
start iteration
done sorting
end inner loop
Iteration 4, inertia 384.4020421577234
start iteration
done sorting
end inner loop
Iteration 5, inertia 383.08768776110526
start iteration
done sorting
end inner loop
Iteration 6, inertia 381.6187216144523
start iteration
done sorting
end inner loop
Iteration 7, inertia 379.90775914845716
start iteration
done sorting
end inner loop
Iteration 8, inertia 378.88393797757396
start iteration
done sorting
end inner loop
Iteration 9, inertia 377.94375998503006
start iteration
done sorting
end inner loop
Iteration 10, inertia 377.2474367656432
start iteration
done sorting
end inner loop
Iteration 11

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=500,
    n_clusters=7, n_init=10, n_jobs=None, precompute_distances='auto',
    random_state=None, tol=0.0001, verbose=1)

In [19]:
print (Features_list)

['REGION-CENTROID-COL', 'REGION-CENTROID-ROW', 'REGION-PIXEL-COUNT', 'SHORT-LINE-DENSITY-5', 'SHORT-LINE-DENSITY-2', 'VEDGE-MEAN', 'VEDGE-SD', 'HEDGE-MEAN', 'HEDGE-SD', 'INTENSITY-MEAN', 'RAWRED-MEAN', 'RAWBLUE-MEAN', 'RAWGREEN-MEAN', 'EXRED-MEAN', 'EXBLUE-MEAN', 'EXGREEN-MEAN', 'VALUE-MEAN', 'SATURATION-MEAN', 'HUE-MEAN']


In [21]:
Kmeans7_class = Kmeans7.labels_
print (Kmeans7_class)
print (Kmeans7.cluster_centers_)

[3 3 3 ... 1 1 0]
[[2.51678367e-01 3.92748756e-01 0.00000000e+00 7.56218884e-02
  1.94029851e-02 7.76573443e-02 4.14943092e-03 6.12403776e-02
  5.03684355e-03 1.47427853e-01 1.37485185e-01 1.84391005e-01
  1.17637066e-01 7.18250172e-01 3.43789293e-01 3.54453556e-01
  1.84648188e-01 4.13413877e-01 2.02752484e-01]
 [7.69062512e-01 4.25930421e-01 0.00000000e+00 1.40237320e-02
  2.26537217e-02 3.97024636e-02 2.98260626e-03 2.31160412e-02
  2.09423436e-03 4.03849749e-02 3.44264190e-02 5.73849797e-02
  2.80592355e-02 7.79917165e-01 2.22794598e-01 4.86886157e-01
  5.83620495e-02 5.39152125e-01 2.44988039e-01]
 [5.35098814e-01 1.50166667e-01 0.00000000e+00 2.77777769e-02
  1.66666667e-03 3.02281387e-02 5.42887957e-04 2.67660451e-02
  5.86661900e-04 8.23246433e-01 7.79716377e-01 8.94170356e-01
  7.88760696e-01 2.70665440e-01 6.66372551e-01 2.89386481e-01
  8.94170356e-01 2.11804171e-01 1.25065773e-01]
 [5.13993692e-01 8.08936588e-01 0.00000000e+00 7.74410751e-02
  5.05050505e-03 5.44737633e-02 

In [22]:
#Transpose to row= freature, column = cluster_centers
Kmeans7_T = Kmeans7.cluster_centers_.T

In [23]:
#Convert the output to dataframe
Kmeans7_table = pd.DataFrame(Kmeans7_T, index=Features_list, columns=['Cluster:0', 'Cluster:1','Cluster:2','Cluster:3','Cluster:4','Cluster:5','Cluster:6'])

In [24]:
Kmeans7_table

Cluster:0  Cluster:1  Cluster:2  Cluster:3  Cluster:4  \
REGION-CENTROID-COL    0.251678   0.769063   0.535099   0.513994   0.748274   
REGION-CENTROID-ROW    0.392749   0.425930   0.150167   0.808937   0.532041   
REGION-PIXEL-COUNT     0.000000   0.000000   0.000000   0.000000   0.000000   
SHORT-LINE-DENSITY-5   0.075622   0.014024   0.027778   0.077441   0.039157   
SHORT-LINE-DENSITY-2   0.019403   0.022654   0.001667   0.005051   0.037651   
VEDGE-MEAN             0.077657   0.039702   0.030228   0.054474   0.113530   
VEDGE-SD               0.004149   0.002983   0.000543   0.001407   0.018922   
HEDGE-MEAN             0.061240   0.023116   0.026766   0.046335   0.107311   
HEDGE-SD               0.005037   0.002094   0.000587   0.001401   0.017627   
INTENSITY-MEAN         0.147428   0.040385   0.823246   0.108790   0.298573   
RAWRED-MEAN            0.137485   0.034426   0.779716   0.091403   0.277521   
RAWBLUE-MEAN           0.184391   0.057385   0.894170   0.092414   0.350081   
RAWGREEN-MEAN          0.117637   0.028059   0.788761   0.142676   0.263837   
EXRED-MEAN             0.718250   0.779917   0.270665   0.679161   0.593300   
EXBLUE-MEAN            0.343789   0.222795   0.666373   0.079002   0.449242   
EXGREEN-MEAN           0.354454   0.486886   0.289386   0.821287   0.311453   
VALUE-MEAN             0.184648   0.058362   0.894170   0.134901   0.350163   
SATURATION-MEAN        0.413414   0.539152   0.211804   0.414491   0.303047   
HUE-MEAN               0.202752   0.244988   0.125066   0.892333   0.164359   

                      Cluster:5  Cluster:6  
REGION-CENTROID-COL    0.302506   0.253603  
REGION-CENTROID-ROW    0.530862   0.459865  
REGION-PIXEL-COUNT     0.000000   0.000000  
SHORT-LINE-DENSITY-5   0.052260   0.026346  
SHORT-LINE-DENSITY-2   0.046610   0.013746  
VEDGE-MEAN             0.100817   0.037337  
VEDGE-SD               0.009420   0.002370  
HEDGE-MEAN             0.083972   0.027901  
HEDGE-SD               0.011043   0.002022  
INTENSITY-MEAN         0.400608   0.025942  
RAWRED-MEAN            0.370347   0.017775  
RAWBLUE-MEAN           0.472461   0.042280  
RAWGREEN-MEAN          0.353036   0.016464  
EXRED-MEAN             0.497146   0.769272  
EXBLUE-MEAN            0.570882   0.216204  
EXGREEN-MEAN           0.213054   0.508343  
VALUE-MEAN             0.472461   0.043251  
SATURATION-MEAN        0.302263   0.802618  
HUE-MEAN               0.163879   0.180506

##### Compare the clustering outout with pre-assigned class

In [25]:
Kmeans7_class

array([3, 3, 3, ..., 1, 1, 0], dtype=int32)

In [26]:
Class_T = np.array(Class.T)[1]
Class_T

array([0, 0, 0, ..., 3, 3, 3], dtype=object)

In [27]:
from sklearn.metrics import completeness_score, homogeneity_score

completeScore = completeness_score(Class_T,Kmeans7_class)
homoScore = homogeneity_score(Class_T,Kmeans7_class)

print ('The completeness of 7 Clusters :  %.8f' %completeScore)
print ('The homogeneity of  7 Clusters :  %.8f' %homoScore)

The completeness of 7 Clusters :  0.61173747
The homogeneity of  7 Clusters :  0.61004999


## c. Perform PCA on the normalized image data matrix. You may use the linear algebra package in Numpy or the Decomposition module in scikit-learn (the latter is much more efficient). Analyze the principal components to determine the number, r, of PCs needed to capture at least 95% of variance in the data. Then use these rcomponents as features to transform the data into a reduced dimension space.

In [28]:
from sklearn import decomposition

In [29]:
#First try n = 5 (capture only 93%)
pca = decomposition.PCA(n_components=5)
IFtrans = pca.fit(IF_norm).transform(IF_norm)
print(pca.explained_variance_ratio_)
print ('The percent of variance in the data  %.8f' %pca.explained_variance_ratio_.sum(0))

[0.60714234 0.13196979 0.10123773 0.04543539 0.03547361]
The percent of variance in the data  0.92125886


In [30]:
#Second try n = 6 (capture only 94%)
pca = decomposition.PCA(n_components=6)
IFtrans = pca.fit(IF_norm).transform(IF_norm)
print(pca.explained_variance_ratio_)
print ('The percent of variance in the data  %.8f' %pca.explained_variance_ratio_.sum(0))

[0.60714234 0.13196979 0.10123773 0.04543539 0.03547361 0.01988035]
The percent of variance in the data  0.94113922


In [31]:
#Third try n = 7 (capture only 96%)
pca = decomposition.PCA(n_components=7)
IFtrans = pca.fit(IF_norm).transform(IF_norm)
print(pca.explained_variance_ratio_)
print ('The percent of variance in the data  %.8f' %pca.explained_variance_ratio_.sum(0))

[0.60714234 0.13196979 0.10123773 0.04543539 0.03547361 0.01988035
 0.0189197 ]
The percent of variance in the data  0.96005892


- At n = 5, it captures only 92%
- At n = 6, it captures 94%
- At n = 7, it captures 96%
**So, we will use n = 7 which captures 96.005892 % of variance in the data.**

#### Transform the data into a reduced dimension space (n = 7)

In [32]:
pca = decomposition.PCA(n_components=7)
IFtrans = pca.fit(IF_norm).transform(IF_norm)
print(IFtrans)

[[-0.68908218  0.53295103  0.24609833 ... -0.07643294  0.0478044
  -0.04732089]
 [-0.66691952  0.51067465  0.33797172 ... -0.04117773  0.05655091
  -0.04170747]
 [-0.71202748  0.77094365 -0.15582189 ... -0.16662228  0.04381397
  -0.06069494]
 ...
 [-0.50774475 -0.12941533 -0.08224551 ... -0.02791769 -0.03189182
  -0.10525024]
 [-0.47909057 -0.08634467 -0.15871141 ...  0.00443463 -0.00371285
  -0.08505898]
 [-0.44225382 -0.10622525 -0.04712725 ...  0.01522668 -0.2061659
   0.15294393]]


In [33]:
pca.explained_variance_ratio_.sum(0)

0.9600589227704959

## d. Perform Kmeans again, but this time on the lower dimensional transformed data. Then, compute the Completeness and Homogeneity values of the new clusters.

In [34]:
from sklearn.cluster import KMeans
Kmeans7_PCA = KMeans(n_clusters=7, max_iter=500,verbose =0)
Kmeans7_PCA.fit(IFtrans)

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=500,
    n_clusters=7, n_init=10, n_jobs=None, precompute_distances='auto',
    random_state=None, tol=0.0001, verbose=0)

In [35]:
Kmeans7_PCA_class = Kmeans7_PCA.labels_
print (Kmeans7_PCA_class)
print (Kmeans7_PCA.cluster_centers_)

[0 0 0 ... 2 2 4]
[[-0.61926803  0.64024955  0.1958285  -0.08685588 -0.06775977  0.00886611
   0.03837202]
 [ 1.41452711  0.0872227   0.03676534 -0.17319537 -0.02992247 -0.00897281
  -0.02157291]
 [-0.51040936 -0.06468741 -0.33636274 -0.06500755  0.0789623   0.0062267
  -0.02632632]
 [ 0.17739221  0.04381971 -0.26494559  0.18454611  0.02677068  0.02440916
   0.00341651]
 [-0.20742333 -0.24636635  0.15253216  0.05661766  0.13031089 -0.00570357
   0.03304526]
 [-0.60370483 -0.35550285  0.10919651 -0.12979947 -0.1309111  -0.02160283
  -0.04388193]
 [ 0.43588833 -0.10529857  0.1652571   0.233499   -0.04503119 -0.00727238
   0.01504941]]


In [36]:
#Transpose to row= freature, column = cluster_centers
Kmeans7_PCA_T = Kmeans7_PCA.cluster_centers_.T
#Convert the output to dataframe
Kmeans7_PCA_table = pd.DataFrame(Kmeans7_PCA_T, index=['PCA1','PCA2','PCA3','PCA4','PCA5','PCA6','PCA7'] , columns=['Cluster:0', 'Cluster:1','Cluster:2','Cluster:3','Cluster:4','Cluster:5','Cluster:6'])

In [37]:
Kmeans7_PCA_table

Cluster:0  Cluster:1  Cluster:2  Cluster:3  Cluster:4  Cluster:5  \
PCA1  -0.619268   1.414527  -0.510409   0.177392  -0.207423  -0.603705   
PCA2   0.640250   0.087223  -0.064687   0.043820  -0.246366  -0.355503   
PCA3   0.195829   0.036765  -0.336363  -0.264946   0.152532   0.109197   
PCA4  -0.086856  -0.173195  -0.065008   0.184546   0.056618  -0.129799   
PCA5  -0.067760  -0.029922   0.078962   0.026771   0.130311  -0.130911   
PCA6   0.008866  -0.008973   0.006227   0.024409  -0.005704  -0.021603   
PCA7   0.038372  -0.021573  -0.026326   0.003417   0.033045  -0.043882   

      Cluster:6  
PCA1   0.435888  
PCA2  -0.105299  
PCA3   0.165257  
PCA4   0.233499  
PCA5  -0.045031  
PCA6  -0.007272  
PCA7   0.015049

In [38]:
Kmeans7_PCA_class

array([0, 0, 0, ..., 2, 2, 4], dtype=int32)

In [39]:
Class_T

array([0, 0, 0, ..., 3, 3, 3], dtype=object)

In [40]:
from sklearn.metrics import completeness_score, homogeneity_score

completeScore = completeness_score(Class_T,Kmeans7_PCA_class)
homoScore = homogeneity_score(Class_T,Kmeans7_PCA_class)

print ('The completeness of 7 Clusters :  %.8f' %completeScore)
print ('The homogeneity of  7 Clusters :  %.8f' %homoScore)

The completeness of 7 Clusters :  0.61157493
The homogeneity of  7 Clusters :  0.60997361


## e. Discuss your observations based on the comparison of the two clustering results.

**Without lower dimensional transformed data (# features = 19)**

- The completeness of 0.61173747 is moderate
- The homogeneity score of 0.61004999 is moderate

**With lower dimensional transformed data (#features using PCA = 7)**

- The completeness of 0.61157493 is moderate
- The homogeneity score of 0.60997361 is moderate
- But both scores are slightly lower than the score of clusturing without lower dimensional transformed data.

The lower dimensional transformed data using PCA when n = 7 can capture 96.005892 % of variance in the data, so their completeness homogeneity scores are similar to the clusturing model without transformation. Comparing to 19 features, the 7 features using PCA can capture the majority of data as well as the 19 features, but the lower dimensionality has cons including faster in processing and running clustering algorithm and better reflecting our intuitions of data. So, in this case using a reduced dimensional data with PCA with the number of features  = 7 would be more suitable for this Kmean clusturing problem. 

## Problem 2

### <font color = red>I use Python 3.7.2 to complete this assignment.</font>
### <font color = red>Some functions in itemBasedRec.py may not compile because the difference in module version
### <font color = red>So, I copied and pasted all functions in this Jupyter Notebook, I also pasted all my new functions in itemBasedRec.py</font>

In [1]:
import os
#Change directory working and import data
os.chdir('/Users/amy_a/Desktop/dsc478/assignment/jokes')

In [2]:
from numpy import *
from numpy import linalg as la
import numpy as np

def ecludSim(inA,inB):
    return 1.0 / (1.0 + la.norm(inA - inB))

def pearsSim(inA,inB):
    if len(inA) < 3 : return 1.0
    return 0.5 + 0.5 * corrcoef(inA, inB, rowvar = 0)[0][1]

def cosSim(inA,inB):
    num = float(inA.T * inB)
    denom = la.norm(inA)*la.norm(inB)
    return 0.5 + 0.5 * (num / denom)

def standEst(dataMat, user, simMeas, item):
    n = shape(dataMat)[1]
    simTotal = 0.0; ratSimTotal = 0.0
    for j in range(n):
        userRating = dataMat[user,j]
        if userRating == 0: continue
        overLap = nonzero(logical_and(dataMat[:,item]>0, \
                                      dataMat[:,j]>0))[0]
        if len(overLap) == 0: similarity = 0
        else: similarity = simMeas(dataMat[overLap,item], \
                                   dataMat[overLap,j])
        #print 'the %d and %d similarity is: %f' % (item, j, similarity)
        simTotal += similarity
        ratSimTotal += similarity * userRating
    if simTotal == 0: return 0
    else: return ratSimTotal/simTotal
    
def svdEst(dataMat, user, simMeas, item):
    n = shape(dataMat)[1]
    simTotal = 0.0; ratSimTotal = 0.0
    data=mat(dataMat)
    U,Sigma,VT = la.svd(data)
    Sig4 = mat(eye(4)*Sigma[:4]) #arrange Sig4 into a diagonal matrix
    xformedItems = data.T * U[:,:4] * Sig4.I  #create transformed items
    for j in range(n):
        userRating = data[user,j]
        if userRating == 0 or j==item: continue
        similarity = simMeas(xformedItems[item,:].T,\
                             xformedItems[j,:].T)
        #print 'the %d and %d similarity is: %f' % (item, j, similarity)
        simTotal += similarity
        ratSimTotal += similarity * userRating
    if simTotal == 0: return 0
    else: return ratSimTotal/simTotal

# This function is not needed for Assignment 4, but may be useful for experimentation
def recommend(dataMat, user, N=3, simMeas=cosSim, estMethod=standEst):
    unratedItems = nonzero(dataMat[user,:].A==0)[1] #find unrated items 
    if len(unratedItems) == 0: return 'you rated everything'
    itemScores = []
    for item in unratedItems:
        estimatedScore = estMethod(dataMat, user, simMeas, item)
        itemScores.append((item, estimatedScore))
    return sorted(itemScores, key=lambda jj: jj[1], reverse=True)[:N]

# This function performs evaluatoin on a single user based on the test_ratio
# For example, with test_ratio = 0.2, a randomly selected 20 percent of rated 
# items by the user are withheld and the rest are used to estimate the withheld ratings

def cross_validate_user(dataMat, user, test_ratio, estMethod=standEst, simMeas=pearsSim):
    number_of_items = np.shape(dataMat)[1]
    rated_items_by_user = np.array([i for i in range(number_of_items) if dataMat[user,i]>0])
    test_size = test_ratio * len(rated_items_by_user)
    test_size = int(test_size)
    test_indices = np.random.randint(0, len(rated_items_by_user), test_size)
    withheld_items = rated_items_by_user[test_indices]
    original_user_profile = np.copy(dataMat[user])
    dataMat[user, withheld_items] = 0 # So that the withheld test items is not used in the rating estimation below
    error_u = 0.0
    count_u = len(withheld_items)

    # Compute absolute error for user u over all test items
    for item in withheld_items:
        # Estimate rating on the withheld item
        estimatedScore = estMethod(dataMat, user, simMeas, item)
        error_u = error_u + abs(estimatedScore - original_user_profile[item])	

    # Now restore ratings of the withheld items to the user profile
    for item in withheld_items:
        dataMat[user, item] = original_user_profile[item]

    # Return sum of absolute errors and the count of test cases for this user
    # Note that these will have to be accumulated for each user to compute MAE
    return error_u, count_u

def test(dataMat, test_ratio, estMethod):
        # Write this function to iterate over all users and for each perform evaluation by calling
        # the above cross_validate_user function on each user. MAE will be the ratio of total error 
        # across all test cases to the total number of test cases, for all users
        totalError = 0
        totalNum = 0
        MAE = 0
        for user in range(dataMat.shape[0]):#iterate over the row
            if estMethod == "svdEst":
                error, count = cross_validate_user(dataMat, user, test_ratio, estMethod=svdEst)
            elif estMethod == "standEst":
                error, count = cross_validate_user(dataMat, user, test_ratio, estMethod=standEst)
            totalError = totalError + error
            totalNum = totalNum + count
        MAE = totalError/totalNum 
        print ('Mean Absoloute Error for ',estMethod,' : ', MAE)

def print_most_similar_jokes(dataMat, jokes, queryJoke, k, metric=pearsSim):
    
    # Write this function to find the k most similar jokes (based on user ratings) to a queryJoke
    # The queryJoke is a joke id as given in the 'jokes.csv' file (an corresponding to the a column in dataMat)
    # You must compare ratings for the queryJoke (the column in dataMat corresponding to the joke), to all
    # other joke rating vectors and return the top k. Note that this is the same as performing KNN on the 
        # columns of dataMat. The function must retrieve the text of the joke from 'jokes.csv' file and print both
    # the queryJoke text as well as the text of the returned jokes.
    
        #Transpose DataMat to (row = user, column = jokes)
        dataMatT = dataMat.T
        numUser = dataMatT.shape[0]
        OneQueryJoke = dataMatT[queryJoke]
    
        totalSim = np.zeros((numUser, 1))#use array to store similarity value for (100 users)
        indexSim = np.zeros((numUser, 1))
        recommendjoke=[]

        #Iterate over each row (user)
        i = 0
        while i < numUser:
            Sim = metric(dataMatT[i],OneQueryJoke)#calulate similarity value
            if Sim ==1:#account for Selected Joke
                totalSim[i]=0
                indexSim[i] = i
            else:
                totalSim[i] = Sim
                indexSim[i] = i
            i += 1
        Results = np.concatenate((totalSim,indexSim), axis =1)
        sortedResults = np.flip(Results[Results[:,0].argsort()])
        indexJoke = sortedResults[:, 0].astype(int)
        for index in indexJoke[:k]:
            recommendjoke.append(index)
    
        print("Selected Joke: \n %s" %jokes[queryJoke])
    
        print ("Top %d Recommended Joke: \n" %k)
        for joke in recommendjoke:
            print(jokes[joke])
            print("________________________________")

def load_jokes(file):
    jokes = genfromtxt(file, delimiter=',', dtype=str)
    jokes = np.array(jokes[:,1])
    return jokes

def get_joke_text(jokes, id):
    return jokes[id]

# dataMat = genfromtxt('modified_jester_data.csv',delimiter=',')

# test(dataMat, 0.2, svdEst)
# test(dataMat, 0.2, standEst)

# jokes = load_jokes('jokes.csv')
# print_most_similar_jokes(dataMat, jokes, 3, 5, pearsSim)


In [3]:
import warnings
warnings.filterwarnings('ignore')

## a. Load in the joke ratings data and the joke text data into appropriate data structures.

In [4]:
import itemBasedRec as IBR

In [5]:
dataJokes = IBR.load_jokes("jokes.csv")
dataJokes.shape

(100,)

In [6]:
dataJokes[0:5]

array(['A man visits the doctor. The doctor says "I have bad news for you.You have cancer and Alzheimer\'s disease". The man replies "Well thank God I don\'t have cancer!"',
       'This couple had an excellent relationship going until one day he came home from work to find his girlfriend packing. He asked her why she was leaving him and she told him that she had heard awful things about him. "What could they possibly have said to make you move out?" "They told me that you were a pedophile." He replied "That\'s an awfully big word for a ten year old."',
       "Q. What's 200 feet long and has 4 teeth? A. The front row at a Willie Nelson Concert.",
       "Q. What's the difference between a man and a toilet? A. A toilet doesn't follow you around after you use it.",
       "Q. What's O. J. Simpson's Internet address? A.\tSlash slash backslash slash slash escape."],
      dtype='<U1198')

In [7]:
from numpy import *
import numpy as np
dataJester = np.genfromtxt("modified_jester_data.csv", delimiter=",")

In [8]:
dataJester

array([[ 3.18, 19.79,  1.34, ...,  0.  ,  0.  ,  0.  ],
       [15.08, 10.71, 17.36, ..., 11.34,  6.68, 12.07],
       [ 0.  ,  0.  ,  0.  , ...,  0.  ,  0.  ,  0.  ],
       ...,
       [16.58, 16.63, 15.85, ...,  0.  ,  0.  ,  0.  ],
       [ 3.67,  4.45,  3.67, ...,  3.77,  3.77,  3.28],
       [ 9.88, 11.73,  9.16, ...,  0.  ,  0.  ,  0.  ]])

## b. Complete the definition for the function "test". This function iterates over all users and for each performs evaluation (by calling the provided "cross_validate_user" function), and returns the error information necessary to compute Mean Absolute Error (MAE). Use this function to perform evaluation (wiht 20% test- ratio for each user) comparing MAE results using standard item-based collaborative filtering (based on the rating prediction function "standEst") with results using the SVD-based version of the rating item-based CF (using "svdEst" as the prediction engine). [Note: See comments provided in the module for hints on accomplishing these tasks.]

## Test function

In [9]:
def test(dataMat, test_ratio, estMethod):
        # Write this function to iterate over all users and for each perform evaluation by calling
        # the above cross_validate_user function on each user. MAE will be the ratio of total error 
        # across all test cases to the total number of test cases, for all users
        totalError = 0
        totalNum = 0
        Mean_Abs_Error = 0
        for user in range(dataMat.shape[0]):#iterate over the row
            if estMethod == "svdEst":
                error, count = cross_validate_user(dataMat, user, test_ratio, estMethod=svdEst)
            elif estMethod == "standEst":
                error, count = cross_validate_user(dataMat, user, test_ratio, estMethod=standEst)
            totalError = totalError + error
            totalNum = totalNum + count
        Mean_Abs_Error = totalError/totalNum 
        print ('Mean Absoloute Error for ',estMethod,' : ', Mean_Abs_Error)

In [10]:
IBR.test(dataJester, 0.2, "standEst")

Mean Absoloute Error for  standEst  :  3.730486391523113


In [11]:
IBR.test(dataJester, 0.2, "svdEst")

Mean Absoloute Error for  svdEst  :  3.663784464984119


- The MAE of from the "standEst" method is slightly higher than the MAE from "svdEst" method.
 
 

## c. Write a new function "print_most_similar_jokes" which takes the joke ratings data, a query joke id, a parameter k for the number of nearest neighbors, and a similarity metric function, and prints the text of the query joke as well as the texts of the top k most similar jokes based on user ratings. [Note: For hints on how to accomplish this task, please see comments at the end of the provided module as well as comments for the provided stub function.]

In [10]:
from numpy import *
import numpy as np
def print_most_similar_jokes(dataMat, jokes, queryJoke, k, metric=pearsSim):
    
    # Write this function to find the k most similar jokes (based on user ratings) to a queryJoke
    # The queryJoke is a joke id as given in the 'jokes.csv' file (an corresponding to the a column in dataMat)
    # You must compare ratings for the queryJoke (the column in dataMat corresponding to the joke), to all
    # other joke rating vectors and return the top k. Note that this is the same as performing KNN on the 
        # columns of dataMat. The function must retrieve the text of the joke from 'jokes.csv' file and print both
    # the queryJoke text as well as the text of the returned jokes.
    
        #Transpose DataMat to (row = user, column = jokes)
        dataMatT = dataMat.T
        numUser = dataMatT.shape[0]
        OneQueryJoke = dataMatT[queryJoke]
    
        totalSim = np.zeros((numUser, 1))#use array to store similarity value for (100 users)
        indexSim = np.zeros((numUser, 1))
        recommendjoke=[]

        #Iterate over each row (user)
        i = 0
        while i < numUser:
            Sim = metric(dataMatT[i],OneQueryJoke)
            if Sim ==1:#account for Selected Joke
                totalSim[i]=0
                indexSim[i] = i
            else:
                totalSim[i] = Sim
                indexSim[i] = i
            i += 1
        Results = np.concatenate((totalSim,indexSim), axis =1)
        sortedResults = np.flip(Results[Results[:,0].argsort()])
        indexJoke = sortedResults[:, 0].astype(int)
        for index in indexJoke[:k]:
            recommendjoke.append(index)
    
        print("Selected Joke: \n %s" %jokes[queryJoke])
    
        print ("Top %d Recommended Joke: \n" %k)
        for joke in recommendjoke:
            print(jokes[joke])
            print("________________________________")

**Test Function with different jokeID ,k, and metric**

In [20]:
IBR.print_most_similar_jokes(dataJester, dataJokes, 3, 5,pearsSim)

Selected Joke: 
 Q. What's the difference between a man and a toilet? A. A toilet doesn't follow you around after you use it.
Top 5 Recommended Joke: 

What do you get when you run over a parakeet with a lawnmower? Shredded tweet.
________________________________
A country guy goes into a city bar that has a dress code and the maitred' demands he wear a tie. Discouraged the guy goes to his car to sulk when inspiration strikes: He's got jumper cables in the trunk! So he wraps them around his neck sort of like a string tie (a bulky string tie to be sure) and returns to the bar. The maitre d' is reluctant but says to the guy "Okay you're a pretty resourceful fellow you can come in... but just don't start anything"!
________________________________
Q. What's 200 feet long and has 4 teeth? A. The front row at a Willie Nelson Concert.
________________________________
What do you call an American in the finals of the world cup?"Hey Beer Man!"
________________________________
Q: What's the dif

In [21]:
IBR.print_most_similar_jokes(dataJester, dataJokes, 3, 5, ecludSim)

Selected Joke: 
 Q. What's the difference between a man and a toilet? A. A toilet doesn't follow you around after you use it.
Top 5 Recommended Joke: 

What do you get when you run over a parakeet with a lawnmower? Shredded tweet.
________________________________
How many teddy bears does it take to change a lightbulb?It takes only one teddybear but it takes a whole lot of lightbulbs.
________________________________
What do you call an American in the finals of the world cup?"Hey Beer Man!"
________________________________
A country guy goes into a city bar that has a dress code and the maitred' demands he wear a tie. Discouraged the guy goes to his car to sulk when inspiration strikes: He's got jumper cables in the trunk! So he wraps them around his neck sort of like a string tie (a bulky string tie to be sure) and returns to the bar. The maitre d' is reluctant but says to the guy "Okay you're a pretty resourceful fellow you can come in... but just don't start anything"!
____________

In [23]:
IBR.print_most_similar_jokes(dataJester, dataJokes, 77, 7,pearsSim)

Selected Joke: 
 Q: What's the difference between the government  and  the Mafia? A: One of them is organized.
Top 7 Recommended Joke: 

If pro- is the opposite of con- then congress must be the opposite of progress.
________________________________
Q: Ever wonder why the IRS calls it Form 1040?A: Because for every $50 that you earn you get 10 and they get 40.
________________________________
On the first day of college the Dean addressed the students pointing out some of the rules:"The female dormitory will be out-of-bounds for all male students and the male dormitory to the female students. Anybody caught breaking this rule will be fined $20 the first time." He continued "Anybody caught breaking this rule the second time will be fined $60. Being caught a third time will cost you a fine of $180. Are there any questions ?"At this point a male student in the crowd inquired:"How much for a season pass ?"
________________________________
Q: How do you keep a computer programmer in the sho